Codec/Archive/Tar/Unpack.hs

```diff
@@ -83,32 +85,59 @@ unpack baseDir entries = do
-               Directory         -> extractDir fn (entryTime entry')
-                                 >> unpackEntries links es'
-               HardLink     link -> (unpackEntries $! saveLink True fn link links) es'
-               SymbolicLink link -> (unpackEntries $! saveLink False fn link links) es'
-               OtherEntryType 'L' _ _ -> throwIO $ userError "Two subsequent OtherEntryType 'L'"
-               _ -> unpackEntries links es'
-             (Fail err)      -> either throwIO throwIO err
-             Done            -> throwIO $ userError "././@LongLink without a subsequent entry"
-      _ -> unpackEntries links es --ignore other file types
+    unpackEntries :: Exception e
+                  => Maybe LinkTarget
+                  -> Maybe FilePath
+                  -> [(FilePath, FilePath, Bool)]     -- ^ links (path, link, isHardLink)
+                  -> Entries (Either e FileNameError) -- ^ entries
+                  -> IO [(FilePath, FilePath, Bool)]
+    unpackEntries _ _ _     (Fail err)      = either throwIO throwIO err
+    unpackEntries _ _ links Done            = return links
+    unpackEntries mLink mPath links (Next entry es) = do
+      let path = fromMaybe (entryPath entry) mPath
+      case entryContent entry of
+        NormalFile file _
+          | Just _ <- mLink -> throwIO $ userError "Expected SymbolicLink or HardLink after OtherEntryType K"
+          | otherwise -> do
+              extractFile (entryPermissions entry) path file (entryTime entry)
+              unpackEntries Nothing Nothing links es
+        Directory
+          | Just _ <- mLink -> throwIO $ userError "Expected SymbolicLink or HardLink after OtherEntryType K"
+          | otherwise -> do
+              extractDir path (entryTime entry)
+              unpackEntries Nothing Nothing links es
+        HardLink     link -> do
+          let linkTarget = fromMaybe link mLink
+          (unpackEntries Nothing Nothing $! saveLink True path linkTarget links) es
+        SymbolicLink link -> do
+          let linkTarget = fromMaybe link mLink
+          (unpackEntries Nothing Nothing $! saveLink False path linkTarget links) es
+        OtherEntryType 'L' fn _
+          | Just _ <- mPath -> throwIO $ userError "Two subsequent OtherEntryType L"
+          | otherwise -> unpackEntries mLink (Just . Char8.unpack . BS.toStrict $ fn) links es
+        OtherEntryType 'K' link _
+          | Just _ <- mLink -> throwIO $ userError "Two subsequent OtherEntryType K"
+          | otherwise -> unpackEntries (Just . LinkTarget . BS.toStrict $ link) mPath links es
+        OtherEntryType _ file _
+          | Just _ <- mLink -> throwIO $ userError "Unknown entry type following OtherEntryType K"
+          | Just _ <- mPath -> throwIO $ userError "Unknown entry type following OtherEntryType L"
+          | otherwise -> do
+              hPutStr stderr "Warning: Unknown tar typecode, attempting to extract as normal file"
```

Comment:
The spec demands that we print to stderr... but since this is a library, it's debatable. Opinions?

Reply:
Let's `throwIO` instead. A library should not write to `stderr` uninvited, it's difficult to catch.

Reply:
That's different semantics than what the spec demands: we don't want to actually error out.

Reply:
Sorry, me being sleepy :) 

Our previous behavior was to unpack unknown entries without a warning/error? Let's stick to it then.

Reply:
> Sorry, me being sleepy :)
> 
> Our previous behavior was to unpack unknown entries without a warning/error? Let's stick to it then.

Previous behavior was to ignore such entries and skip them. It's hard to make an informed decision here without lots of data of odd tar archives, which is why I'd prefer to somewhat stick to the spec.

We could also change the interface of `unpack` (which I think we should, `libarchive` also does that) to something like `unpack :: Exception e => FilePath -> Entries e -> IO (These ArchiveError ArchiveWarning)`.

Exceptions aren't even documented properly.

Reply:
AFAIR between many complaints about this library, none was about unknown entry types. So I'd rather keep it simple. I'm certainly against `stderr`.

The simplest solution: we can stick to the current behaviour and keep skipping such entries. At least, it would not be a regression. If someone comes up with interesting examples of malformed archives in future, we can review the decision.

A more involved solution: change `unpack` to throw an error on unknown entries (which can be caught by client) + provide tools to sanitize `Entries`:
* Add `mapMaybeEntries :: (Entry e -> Maybe (Entry e) -> Entries e -> Entries e`
* Add `isKnownEntry :: Entry e -> Bool`
* Add `replaceUnknownEntries :: Entry e -> Entry e`, which replaces unknown `OtherEntryType` with `NormalFile`.